rust/src/server.rs

```diff
@@ -0,0 +1,190 @@
+            .match_header("content-type", "application/test")
+            .match_header("DD-API-KEY", "SECRET API KEY")
+            .match_header("user-agent", "unit-test")
+            .with_body("{expected response}")
+            .with_status(418)
+            .expect(1)
+            .create();
+        let server = TestServer::serve(server_address());
+        let uri_string = format!("http://{}/some/test/path", server.addr);
+        let uri = uri_string
+            .parse::<hyper::Uri>()
+            .expect("server addr should parse");
+        let client = Client::new();
+        let req = Request::builder()
+            .method(Method::POST)
+            .header("content-type", "application/test")
+            .header("DD-API-KEY", "SECRET API KEY")
+            .header("user-agent", "unit-test")
+            .uri(uri)
+            .body(Body::from("{expected payload}"))
+            .expect("request builder");
+        let resp = client.request(req).await.unwrap();
+        assert_eq!(resp.status(), 418);
+        mock.assert();
+    }
+
+    pub fn runtime() -> tokio::runtime::Runtime {
+        tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("new rt")
+    }
+
+    fn tcp_bind(addr: &SocketAddr) -> ::tokio::io::Result<TcpListener> {
+        let std_listener = StdTcpListener::bind(addr).unwrap();
+        std_listener.set_nonblocking(true).unwrap();
+        TcpListener::from_std(std_listener)
+    }
+
+    struct TestServer {
+        shutdown_signal: Option<oneshot::Sender<()>>,
+        thread: Option<thread::JoinHandle<()>>,
+        addr: SocketAddr,
+    }
+
+    impl TestServer {
+        fn serve(proxy_addr: SocketAddr) -> TestServer {
+            let (shutdown_tx, shutdown_rx) = oneshot::channel();
+            let listener = tcp_bind(&"127.0.0.1:0".parse().unwrap()).unwrap();
+            let addr = listener.local_addr().unwrap();
+            let thread = thread::Builder::new()
+                .spawn(move || {
+                    runtime()
+                        .block_on(async move {
+                            let proxy_client = ProxyClient::new(addr, proxy_addr);
+                            let server = new!(proxy_client);
+                            server
+                                .with_graceful_shutdown(async {
+                                    let _ = shutdown_rx.await;
+                                })
+                                .await
+                        })
+                        .expect("serve()");
+                })
+                .expect("thread spawn");
+            return TestServer {
+                shutdown_signal: Some(shutdown_tx),
+                thread: Some(thread),
+                addr,
+            };
+        }
+    }
+
+    impl Drop for TestServer {
+        fn drop(&mut self) {
+            drop(self.shutdown_signal.take());
+            drop(self.thread.take());
+        }
```

Comment:
This seems to be unnecessary -- removing it doesn't change the behavior.  The default drop implementation for a struct drops the struct fields (in order, but I don't think that matters here?)